lib/cn/spill.c

```diff
@@ -180,394 +180,437 @@ kv_spill_abort(struct cn_tstate_omf *omf, void *arg)
+ */
+merr_t
+cn_spill(struct cn_compaction_work *w)
+{
+    struct bin_heap *bh;
+    struct merge_item curr;
+    struct kvset_builder *child = NULL;
+    struct cn_khashmap *khashmap = NULL;
+    struct key_obj ekobj = { 0 }, prev_kobj;
+    struct cn_tree *tree = w->cw_tree;
 
-            if (!buf || bufsz < bufsz_min) {
-                free_aligned(buf);
+    char ekbuf[HSE_KVS_KEY_LEN_MAX];
+    uint vlen, complen, omlen, eklen, direct_read_len, cnum, prev_cnum;
+    uint curr_klen HSE_MAYBE_UNUSED;
+    u32 bufsz = 0;
+    void *buf = NULL;
+    merr_t err;
 
-                if (omlen < HSE_KVS_VALUE_LEN_MAX / 4)
-                    bufsz = HSE_KVS_VALUE_LEN_MAX / 4;
-                else if (omlen < HSE_KVS_VALUE_LEN_MAX / 2)
-                    bufsz = HSE_KVS_VALUE_LEN_MAX / 2;
-                else
-                    bufsz = HSE_KVS_VALUE_LEN_MAX;
+    u64  seq, emitted_seq = 0, emitted_seq_pt = 0;
+    bool emitted_val = false, bg_val = false, more, gt_max_edge = false;
+    bool kvcompact = (w->cw_action == CN_ACTION_COMPACT_KV);
+    bool spill = (w->cw_action == CN_ACTION_SPILL);
 
-                /* add an extra page if not aligned */
-                if (bufsz_min < bufsz)
-                    bufsz += PAGE_SIZE;
+    u64  tstart, tprog = 0;
+    u64  dbg_prev_seq HSE_MAYBE_UNUSED;
+    uint dbg_prev_src HSE_MAYBE_UNUSED;
+    uint dbg_nvals_this_key HSE_MAYBE_UNUSED;
+    uint seqno_errcnt = 0;
+    bool dbg_dup HSE_MAYBE_UNUSED;
 
-                buf = alloc_aligned(bufsz, PAGE_SIZE);
-                if (!buf) {
-                    err = merr(ENOMEM);
-                    goto done;
-                }
+    /* Variables for tracking the last ptomb context */
+    struct key_obj pt_kobj = { 0 };
+    bool pt_set = false;
+    u64 pt_seq = 0;
+
+    /* TODO: Using spill by hash will be gone once we completely move to a 2-level cN tree
+     * with a full fledged route map
+     */
+    if (!tree->rp->cn_incr_rspill || !tree->ct_route_map || (spill && w->cw_level > 0))
+        return cn_spill_hash(w);
+
+    assert(w->cw_kvset_cnt);
+    assert(w->cw_inputv);
+
+    if (w->cw_prog_interval && w->cw_progress)
+        tprog = jiffies;
+
+    /* We must issue a direct read for all values that will not fit into the vblock readahead
+     * buffer.  Since all direct reads require page size alignment any value whose length is
+     * greater than the buffer size minus one page must be read directly from disk (vs from the
+     * readahead buffer).
+     */
+    direct_read_len = w->cw_rp->cn_compact_vblk_ra;
+    direct_read_len -= PAGE_SIZE;
+
+    memset(w->cw_outv, 0, w->cw_outc * sizeof(*w->cw_outv));
+
+    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    if (err)
+        return err;
+
+    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
+    if (more)
+        cnum = get_route(w, &curr.kobj, ekbuf, sizeof(ekbuf), &eklen);
```

Comment:
Add a todo for eliminating the need to call get_route on every key.

Reply:
This is already done in this version. get_route() is called only when the current key exceeds the edge key. It is not called for every key. Please check the while loop at line 405.

Reply:
There are three nested while loops - the first one loops across nodes, the second one loops across keys within a node and the third one loops across values within a key. get_route() is called only in the outermost loop to determine the next child node number and its edge key.